### PR TITLE
[CARBONDATA-777] Alter table support for spark 2.1

### PR DESCRIPTION
**Pull request by manishgupta88**
As part of this PR following features will be supported.

1. Support alter table result preparation.
2. Support reading data with different block key generators.
3. Support addition of a new column.
4. Support deletion of a column.
5. Support change in data type form INT to BIGINT
6. Support Change of decimal datatype from lower to higher precision.
7. Support filtering on newly added columns.
8. Support rename table
9. Parsing support for the new DDL commands added.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1068/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1071/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1072/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1073/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1076/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1083/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1102/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1105/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1110/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1113/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1114/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1115/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1116/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1126/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1129/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 1]
```diff
@@ -966,20 +974,22 @@ public static void clearBlockCache(List<AbstractIndex> dataBlocks) {
...
-      //assuming complex dimensions will always be atlast
-      if (null != childs && childs.size() > 0) {
-        break;
-      }
-      if (carbonDimension.isColumnar() && hasEncoding(carbonDimension.getEncoder(),
-          Encoding.DICTIONARY)) {
-        isDictionaryDimensions.add(true);
-      } else if (!carbonDimension.isColumnar()) {
-        if (processedColumnGroup.add(carbonDimension.columnGroupId())) {
+      if (!carbonDimension.isInvisible()) {
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: The list which is being created, should have two getters. one to get With invisible and other to get without invisible. Use required function accordingly

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 2]
```diff
@@ -1520,6 +1572,69 @@ public static String getFormatFromProperty(DataType dataType) {
   }
 
   /**
+   * This method will delete the dictionary files for the given column IDs and
+   * clear the dictionary cache
+   *
+   * @param dictionaryColumns
+   * @param carbonTable
+   */
+  public static void deleteDictionaryFileAndCache(List<CarbonColumn> dictionaryColumns,
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: Create a class to manage Dictionary and move this method into that

File: core/src/main/java/org/apache/carbondata/core/util/DataTypeUtil.java [hunk 3]
```diff
@@ -245,6 +248,36 @@ public static DataType getDataType(String dataTypeStr) {
...
 
   /**
+   * This method will convert the data according to its data type and perform a
+   * special handling for decimal data types
+   *
+   * @param dataInBytes
+   * @param dimension
+   * @return
+   */
+  public static Object getDataBasedOnDataType(byte[] dataInBytes, CarbonDimension dimension) {
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/util/DataTypeUtil.java]: change the previous function to handle the logic
 getDataBasedOnDataType(dataInBytes, dimension.getDataType());

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala [hunk 4]
```diff
@@ -129,4 +134,80 @@ class CarbonSpark2SqlParser extends CarbonDDLSqlParser {
       case databaseName ~ tableName ~ limit =>
         ShowLoadsCommand(convertDbNameToLowerCase(databaseName), tableName.toLowerCase(), limit)
     }
+
+  protected lazy val alterTableModifyDataType: Parser[LogicalPlan] =
+    ALTER ~> TABLE ~> (ident <~ ".").? ~ ident ~ CHANGE ~ ident ~ ident ~
+    ident ~ opt("(" ~> rep1sep(valueOptions, ",") <~ ")") <~ opt(";") ^^ {
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: Use datatype rule of existing parser rules, I think rule name is PrimitiveType.

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala [hunk 5]
```diff
@@ -129,4 +134,80 @@ class CarbonSpark2SqlParser extends CarbonDDLSqlParser {
...
+      case dbName ~ table ~ change ~ columnName ~ columnNameCopy ~ dataType ~ values =>
+        // both the column names should be same
+        CommonUtil.validateColumnNames(columnName, columnNameCopy)
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: Directly check equalsIgnoreCase, no separate util function required

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala (hunk 4 above)
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: Move this to SparkDDLParser

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala [hunk 6]
```diff
@@ -129,4 +134,80 @@ class CarbonSpark2SqlParser extends CarbonDDLSqlParser {
...
+          tableModel.msrCols,
+          tableModel.highcardinalitydims.getOrElse(Seq.empty))
+        AlterTableAddColumns(alterTableAddColumnsModel)
+    }
+
+  private def convertFieldNamesToLowercase(field: Field): Field = {
+    val name = field.column.toLowerCase
+    field.copy(column = name, name = Some(name))
+  }
+  protected lazy val alterTableDropColumn: Parser[LogicalPlan] =
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: Add rule name columnlist, which can check duplicates and related validation

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetastore.scala [hunk 7]
```diff
@@ -304,38 +341,76 @@ class CarbonMetastore(conf: RuntimeConfig, val storePath: String) {
...
     val fileType = FileFactory.getFileType(schemaMetadataPath)
     if (!FileFactory.isFileExist(schemaMetadataPath, fileType)) {
       FileFactory.mkdirs(schemaMetadataPath, fileType)
     }
     val thriftWriter = new ThriftWriter(schemaFilePath, false)
-    thriftWriter.open()
+    thriftWriter.open(FileWriteOperation.OVERWRITE)
     thriftWriter.write(thriftTableInfo)
     thriftWriter.close()
+    removeTableFromMetadata(dbName, tableName)
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetastore.scala]: directly call refresh flow after updating alter schema

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 8]
```diff
@@ -136,6 +140,298 @@ case class AlterTableCompaction(alterTableModel: AlterTableModel) extends Runnab
   }
 }
 
+private[sql] case class AlterTableDataTypeChange(
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: Alter table related move to AlterTableCommands file

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 9]
```diff
@@ -136,6 +140,298 @@ case class AlterTableCompaction(alterTableModel: AlterTableModel) extends Runnab
...
+      .schema.SchemaEvolutionEntry()
+      schemaEvolutionEntry.setTimeStamp(System.currentTimeMillis)
+      schemaEvolutionEntry.setAdded(newCols.toList.asJava)
+
+      val thriftTable = schemaConverter
+        .fromWrapperToExternalTableInfo(wrapperTableInfo, dbName, tableName)
+      thriftTable.getFact_table.getSchema_evolution.getSchema_evolution_history.get(0)
+        .setTime_stamp(System.currentTimeMillis)
+      CarbonEnv.get.carbonMetastore
+        .updateTableSchema(carbonTable.getCarbonTableIdentifier,
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: Write common functions, avoid duplicate code.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 10]
```diff
@@ -136,6 +140,298 @@ case class AlterTableCompaction(alterTableModel: AlterTableModel) extends Runnab
...
+            deletedColumnSchema += CarbonScalaUtil.createColumnSchemaCopyObject(columnSchema)
+            columnSchema.invisible = true
+          }
+        }
+      }
+      // add deleted columns to schema evolution history and update the schema
+      tableInfo.getFact_table.getSchema_evolution.getSchema_evolution_history.get(0)
+        .setTime_stamp(System.currentTimeMillis)
+      val schemaEvolutionEntry = new SchemaEvolutionEntry(System.currentTimeMillis)
+      schemaEvolutionEntry.setRemoved(deletedColumnSchema.toList.asJava)
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: Add function addEvolutionEntry

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1147/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/DataTypeConverterUtil.scala [hunk 11]
```diff
@@ -77,4 +77,40 @@ object DataTypeConverterUtil {
...
     }
   }
+
+  /**
+   * convert from wrapper to external data type
+   *
+   * @param dataType
+   * @return
+   */
+  def convertToThriftDataType(dataType: String): org.apache.carbondata.format.DataType = {
```
Comment (nareshpr) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/DataTypeConverterUtil.scala]: There is no direct string to thrift datatype conversion. When creating table, we first convert string to Wrapper datatype and from wrapper datatype, we convert into thrift datatype. This method is required for alter table, as we r directly converting string to thrift type in alter table change datatype

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala [hunk 12]
```diff
@@ -194,4 +196,102 @@ object CarbonScalaUtil {
...
+    }
+  }
+
+  /**
+   * This method will create a copy of the same object
+   *
+   * @param thriftColumnSchema object to be cloned
+   * @return
+   */
+  def createColumnSchemaCopyObject(thriftColumnSchema: org.apache.carbondata.format.ColumnSchema)
```
Comment (nareshpr) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala]: Done

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetastore.scala (hunk 7 above)
Comment (nareshpr) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetastore.scala]: As per old code, when new table is created, table info is added into CarbonMetastore and modified.mdt file is timestamp is updated to refresh in other sessions. Those changes are extracted to a new method and used in CreateTable and AlterTable flow, updating mdt file code is missing, which i added

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala (hunk 5 above)
Comment (nareshpr) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: Ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala (hunk 6 above)
Comment (nareshpr) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: Ok

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1154/

**Comment by manishgupta88** [comment]
@gvramana ...Handled all review comments and all the test cases are passing. Please refer the below link.
kindly review and merge

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1156/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1160/

**Comment by nareshpr** [comment]
@gvramana ...Handled all review comments and all the test cases are passing. Please refer the below link.
http://136.243.101.176:8080/job/ManualApacheCarbonPRBuilder2.1/103/
kindly review and merge

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1162/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 13]
```diff
@@ -747,14 +835,14 @@ case class LoadTable(
             true
           } else {
             LOGGER.error("Can't use single_pass, because SINGLE_PASS and ALL_DICTIONARY_PATH" +
-              "can not be used together, and USE_KETTLE must be set as false")
+                         "can not be used together, and USE_KETTLE must be set as false")
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: wrong indentation

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 14]
```diff
@@ -313,6 +307,100 @@ private[sql] case class AlterTableAddColumns(
   }
 }
 
+private[sql] case class AlterTableRenameTable(alterTableRenameModel: AlterTableRenameModel)
```
Comment (nareshpr) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: Done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala (hunk 13 above)
Comment (nareshpr) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: Done

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1163/

**Comment by gvramana** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1164/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1165/

**Comment by gvramana** [comment]
compilation errors in 1.6.2, please fix the same

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1166/

**Comment by gvramana** [comment]
modified ManualApacheCarbonPRBuilder2.1 to test spark1.6 and tried building. tests have failed, @manishgupta88 please fix.
http://136.243.101.176:8080/job/ManualApacheCarbonPRBuilder2.1/107/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1178/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1182/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1183/

**Comment by manishgupta88** [comment]
@gvramana ...Fixed test cases failure in spark1.6.2. Please refer the below link
http://136.243.101.176:8080/job/ManualApacheCarbonPRBuilder2.1/112/

For spark-2.1 --> http://136.243.101.176:8080/job/ManualApacheCarbonPRBuilder2.1/111/

**Pull request closed without merging**